leoism/step/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,124 @@
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> slotsOpen = new ArrayList<TimeRange>();
+
+    // A requested meeting duration longer than the time of day should not be possible.
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return slotsOpen;
+    }
+
+    if (events.isEmpty()) {
+      slotsOpen.add(TimeRange.WHOLE_DAY);
+      return slotsOpen;
+    }
+
+    ArrayList<TimeRange> slotsTaken = getSlotsTaken(events, request);
+
+    int previousEndTime = TimeRange.START_OF_DAY;
+    for (TimeRange occupied : slotsTaken) {
+      slotsOpen.add(TimeRange.fromStartEnd(previousEndTime, occupied.start(), false));
+      previousEndTime = occupied.end();
+    }
+
+    slotsOpen.add(TimeRange.fromStartEnd(previousEndTime, TimeRange.END_OF_DAY, true));
+
+    return sufficientTimeAvailable(slotsOpen, request);
+  }
+
+  /**
+   * Checks if the requested time interval overlaps with another. If true, then the conflicting
+   * TimeRange is removed and a new TimeRange is returned with the events time merged. Otherwise,
+   * returns null if no TimeRange overlaps.
+   */
+  private TimeRange doesOverlap(TimeRange requestedTime, ArrayList<TimeRange> slotsTaken) {
+    for (TimeRange occupied : slotsTaken) {
+      if (requestedTime.overlaps(occupied)) {
+        // If an overlap occurs, merge them by getting the earliest start time and the latest start
+        // time.
+        int startTime =
+            occupied.start() > requestedTime.start() ? requestedTime.start() : occupied.start();
+        int endTime = occupied.end() > requestedTime.end() ? occupied.end() : requestedTime.end();
+        slotsTaken.remove(occupied);
+        return TimeRange.fromStartEnd(startTime, endTime, false);
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Checks if all required attendees (those that the event is booked under) are the same as the one
+   * making the Meeting Request. Returns false if a required attendee is not present. Otherwise,
+   * returns true.
+   */
+  private boolean areRequiredAttendeesPresent(
+      Collection<String> requiredAttendees, Collection<String> presentAttendees) {
+    for (String attendee : presentAttendees) {
+      if (!requiredAttendees.contains(attendee)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Stores all attendees which the event is booked under and stores them inside a Set to avoid
+   * duplicates. Returns a HashSet with all attendees booked under an event.
+   */
+  private HashSet<String> allEventAttendees(Collection<Event> events) {
+    HashSet<String> eventAttendees = new HashSet<String>();
+    for (Event event : events) {
+      eventAttendees.addAll(event.getAttendees());
+    }
+
+    return eventAttendees;
+  }
+
+  /** Returns an ArrayList of TimeRange of all slots that are occupied by events. */
+  private ArrayList<TimeRange> getSlotsTaken(Collection<Event> events, MeetingRequest request) {
```

Comment:
Feels like a better name would be like "getCombinedSlotsTaken" or something like taht.